File: frontend/my-autoparts/src/utils/productDisplayName.js
function escapeRegex(value) {
  return String(value).replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function normalizeArticle(value) {
  if (!value) return '';
  return String(value).replace(/[^A-Za-z0-9А-Яа-яЁё]/gi, '').toUpperCase();
}

function stripLeadingArticle(text, article) {
  let result = String(text || '').trim();
  const articleStr = String(article || '').trim();
  if (!result || !articleStr) return result;

  const articleNorm = normalizeArticle(articleStr);
  const exactRe = new RegExp(`^${escapeRegex(articleStr)}(?:\\s+|$)`, 'i');
  result = result.replace(exactRe, '').trim();

  const firstToken = result.split(/\s+/)[0] || '';
  if (articleNorm && normalizeArticle(firstToken) === articleNorm) {
    result = result.split(/\s+/).slice(1).join(' ').trim();
  }

  return result;
}

function stripLeadingBrand(text, brand) {
  const brandStr = String(brand || '').trim();
  if (!text || !brandStr) return String(text || '').trim();

  const brandRe = new RegExp(`^${escapeRegex(brandStr)}\\s+`, 'i');
  return String(text).replace(brandRe, '').trim();
}

/**
 * Извлекает описание без категории, бренда и артикула (формат Rossko: «Категория BRAND / ARTICLE Описание»).
 */
export function extractProductDescription(rawName, brand, article) {
  let name = String(rawName || '').trim();
  if (!name) return '';

  const brandStr = String(brand || '').trim();
  const articleStr = String(article || '').trim();

  const slashParts = name.split('/').map((part) => part.trim()).filter(Boolean);
  if (slashParts.length >= 2) {
    const afterSlash = slashParts.slice(1).join(' / ').trim();
    const words = afterSlash.split(/\s+/).filter(Boolean);
    if (words.length > 1) {
      const withoutLeadingArticle = stripLeadingArticle(afterSlash, articleStr);
      if (withoutLeadingArticle) return withoutLeadingArticle;
      return words.slice(1).join(' ').trim();
    }
    return stripLeadingArticle(afterSlash, articleStr);
  }

  if (brandStr) {
    const brandSuffixRe = new RegExp(`^(?:.+?\\s+)?${escapeRegex(brandStr)}\\s*`, 'i');
    name = name.replace(brandSuffixRe, '').trim();
  }

  name = stripLeadingArticle(name, articleStr);
  name = stripLeadingBrand(name, brandStr);
  name = stripLeadingArticle(name, articleStr);

  return name.trim();
}

/**
 * Заголовок товара: «Бренд Артикул Название».
 */
export function formatProductDisplayTitle(brand, article, rawName) {
  const brandStr = String(brand || '').trim();
  const articleStr = String(article || '').trim();
  const raw = String(rawName || '').trim();

  if (!brandStr && !articleStr) {
    return raw || '—';
  }

  const description =
    extractProductDescription(raw, brandStr, articleStr) || raw;

  const parts = [brandStr, articleStr, description].filter(Boolean);
  const formatted = parts.join(' ').replace(/\s+/g, ' ').trim();

  return formatted || '—';
}

File: backend/app/utils/product_display_name.py
from __future__ import annotations

import re


def _escape_regex(value: str) -> str:
    return re.escape(str(value or ""))


def _normalize_article(value: str | None) -> str:
    if not value:
        return ""
    return re.sub(r"[^A-Za-z0-9А-Яа-яЁё]", "", str(value), flags=re.IGNORECASE).upper()


def _strip_leading_article(text: str, article: str | None) -> str:
    result = str(text or "").strip()
    article_str = str(article or "").strip()
    if not result or not article_str:
        return result

    article_norm = _normalize_article(article_str)
    exact_re = re.compile(rf"^{_escape_regex(article_str)}(?:\s+|$)", re.IGNORECASE)
    result = exact_re.sub("", result).strip()

    first_token = (result.split() or [""])[0]
    if article_norm and _normalize_article(first_token) == article_norm:
        result = " ".join(result.split()[1:]).strip()

    return result


def _strip_leading_brand(text: str, brand: str | None) -> str:
    brand_str = str(brand or "").strip()
    if not text or not brand_str:
        return str(text or "").strip()
    brand_re = re.compile(rf"^{_escape_regex(brand_str)}\s+", re.IGNORECASE)
    return brand_re.sub("", str(text)).strip()


def extract_product_description(raw_name: str | None, brand: str | None, article: str | None) -> str:
    name = str(raw_name or "").strip()
    if not name:
        return ""

    brand_str = str(brand or "").strip()
    article_str = str(article or "").strip()

    slash_parts = [part.strip() for part in name.split("/") if part.strip()]
    if len(slash_parts) >= 2:
        after_slash = " / ".join(slash_parts[1:]).strip()
        words = after_slash.split()
        if len(words) > 1:
            without_article = _strip_leading_article(after_slash, article_str)
            if without_article:
                return without_article
            return " ".join(words[1:]).strip()
        return _strip_leading_article(after_slash, article_str)

    if brand_str:
        brand_suffix_re = re.compile(rf"^(?:.+?\s+)?{_escape_regex(brand_str)}\s*", re.IGNORECASE)
        name = brand_suffix_re.sub("", name).strip()

    name = _strip_leading_article(name, article_str)
    name = _strip_leading_brand(name, brand_str)
    name = _strip_leading_article(name, article_str)
    return name.strip()


def format_product_display_title(
    brand: str | None,
    article: str | None,
    raw_name: str | None,
) -> str:
    brand_str = str(brand or "").strip()
    article_str = str(article or "").strip()
    raw = str(raw_name or "").strip()

    if not brand_str and not article_str:
        return raw or "Автозапчасть"

    description = extract_product_description(raw, brand_str, article_str) or raw
    parts = [brand_str, article_str, description]
    formatted = " ".join(part for part in parts if part).strip()
    return formatted or "Автозапчасть"
